configs/provider_validation.go

```diff
@@ -0,0 +1,243 @@
+			Subject:  &call.DeclRange,
+		})
+	}
+
+	// You cannot pass in a provider that cannot be used
+	for name, passed := range passedIn {
+		providerAddr := addrs.AbsProviderConfig{
+			Module:   cfg.Path,
+			Provider: addrs.NewDefaultProvider(passed.InChild.Name),
+			Alias:    passed.InChild.Alias,
+		}
+
+		localAddr, localName := localNames[name]
+		if localName {
+			providerAddr = localAddr
+		}
+
+		aliasAddr, configAlias := configAliases[name]
+		if configAlias {
+			providerAddr = aliasAddr
+		}
+
+		_, emptyConfig := emptyConfigs[name]
+
+		if !(localName || configAlias || emptyConfig) {
+			severity := hcl.DiagError
+
+			// we still allow default configs, so switch to a warning if the incoming provider is a default
+			if providerAddr.Provider.IsDefault() {
+				severity = hcl.DiagWarning
+			}
+
+			diags = append(diags, &hcl.Diagnostic{
+				Severity: severity,
+				Summary:  fmt.Sprintf("Provider %s is undefined", name),
+				Detail: fmt.Sprintf("Module %s does not declare a provider named %s.\n", cfg.Path, name) +
+					fmt.Sprintf("If you wish to specify a provider configuration for the module, add an entry for %s in the required_providers block within the module.", name),
+				Subject: &passed.InChild.NameRange,
+			})
+		}
+
+		// The provider being passed in must also be of the correct type.
+		// While we would like to ensure required_providers exists here,
+		// implied default configuration is still allowed.
+		pTy := addrs.NewDefaultProvider(passed.InParent.Name)
+
+		// use the full address for a nice diagnostic output
+		parentAddr := addrs.AbsProviderConfig{
+			Module:   cfg.Parent.Path,
+			Provider: pTy,
+			Alias:    passed.InParent.Alias,
+		}
+
+		if cfg.Parent.Module.ProviderRequirements != nil {
+			req, defined := cfg.Parent.Module.ProviderRequirements.RequiredProviders[name]
+			if defined {
+				parentAddr.Provider = req.Type
+			}
+		}
+
+		if !providerAddr.Provider.Equals(parentAddr.Provider) {
+			diags = append(diags, &hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  fmt.Sprintf("Invalid type for provider %s", providerAddr),
+				Detail: fmt.Sprintf("Cannot use configuration from %s for %s. ", parentAddr, providerAddr) +
+					"The given provider configuration is for a different provider type.",
+				Subject: &passed.InChild.NameRange,
+			})
+		}
+	}
+
+	// Empty configurations are no longer needed
+	for name, src := range emptyConfigs {
+		detail := fmt.Sprintf("Remove the %s provider block from %s.", name, cfg.Path)
+
+		isAlias := strings.Contains(name, ".")
+		_, isConfigAlias := configAliases[name]
+		_, isLocalName := localNames[name]
+
+		if isAlias && !isConfigAlias {
+			localName := strings.Split(name, ".")[0]
+			detail = fmt.Sprintf("Remove the %s provider block from %s. Add %s to the list of configuration_aliases for %s in required_providers to define the provider configuration name.", name, cfg.Path, name, localName)
```

Comment:
I worry a bit that users might not find this easy to understand, given that we're introducing `configuration_aliases` in the same release we're adding these warnings.

Assuming that we're not going to provide an auto-upgrade command, could it make sense to make this warning even _more_ explicit about what to do? Maybe with sample configuration?

Reply:
Yes, I was thinking the same thing! I'm going to see how this gets formatted with config in the `detail` section. My hesitation was that I can't always generate the complete `required_providers` attribute, and it could cause more confusion than if they just look up the docs 🤔

Reply:
ok, adding config examples is turning into a much larger task. I'll fix the dandling comment, and get this merged. Other improvements can come later.